source/opt/constructs.h

```diff
@@ -0,0 +1,305 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+// This file defines the language constructs for representing a SPIR-V
+// module in memory.
+
+#ifndef LIBSPIRV_OPT_CONSTRUCTS_H_
+#define LIBSPIRV_OPT_CONSTRUCTS_H_
+
+#include <functional>
+#include <vector>
+
+#include "spirv-tools/libspirv.h"
+#include "spirv/1.1/spirv.h"
+
+namespace spvtools {
+namespace opt {
+namespace ir {
+
+class Function;
+class Module;
+
+// About operand:
+//
+// In the SPIR-V specification, the term "operand" is used to mean any single
+// SPIR-V word following the leading wordcount-opcode word. Here, the term
+// "operand" is used to mean a *logical* operand. A logical operand may consist
+// of mulitple SPIR-V words, which together make up the same component. For
+// example, a logical operand of a 64-bit integer needs two words to express.
+//
+// Further, we categorize logical operands into *in* and *out* operands.
+// In operands are operands actually serve as input to operations, while out
+// operands are operands that represent ids generated from operations (result
+// type id or result id). For example, for "OpIAdd %rtype %rid %inop1 %inop2",
+// "%inop1" and "%inop2" are in operands, while "%rtype" and "%rid" are out
+// operands.
+
+// A *logical* operand to a SPIR-V instruction. It can be the type id, result
+// id, or other additional operands carried in an instruction.
+struct Operand {
+  Operand(spv_operand_type_t t, std::vector<uint32_t>&& w)
+      : type(t), words(std::move(w)) {}
```

Comment:
Oh wait, if the actual argument is an lvalue reference, then by reference collapsing rules, then w itself will be an lvalue reference. In that case you do want the std::move.